src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp

```diff
@@ -0,0 +1,387 @@
+/// `double OP DataVector = DataVector`.
+///
+/// To enable binary operations between two types that is not yet supported,
+/// modify a current template specialization or add a new one.
+///
+/// \tparam X1 the data type of one operand
+/// \tparam X2 the data type of the other operand
+template <typename X1, typename X2>
+struct get_binop_datatype_impl;
+
+/// No template specialization was matched, so it's not a known pairing
+template <typename X1, typename X2>
+struct get_binop_datatype_impl {
+  using type = std::bool_constant<false>;
+};
+/// A binary operation between two terms of the same type will yield a result
+/// with that type
+template <typename X>
+struct get_binop_datatype_impl<X, X> {
+  using type = X;
+};
+/// A binary operation between two `VectorImpl`s of the same type will
+/// yield the shared derived `VectorImpl`, e.g.
+/// `DataVector OP DataVector = DataVector`
+template <typename ValueType, typename VectorType>
+struct get_binop_datatype_impl<VectorImpl<ValueType, VectorType>,
+                               VectorImpl<ValueType, VectorType>> {
+  using type = VectorType;
+};
+/// @{
+/// A binary operation between a `VectorImpl` and its underlying value type
+/// yields the `VectorImpl`, e.g. `DataVector OP double = DataVector`
+template <typename ValueType, typename VectorType>
+struct get_binop_datatype_impl<VectorImpl<ValueType, VectorType>, ValueType> {
+  using type = VectorType;
+};
+template <typename ValueType, typename VectorType>
+struct get_binop_datatype_impl<ValueType, VectorImpl<ValueType, VectorType>> {
+  using type = VectorType;
+};
+/// @}
+
+/// \brief Get the data type of a binary operation between two data types
+/// that may occur in a `TensorExpression`
+///
+/// \details
+/// See `get_binop_datatype_impl` for which data type combinations have a
+/// defined result type
+///
+/// \tparam X1 the data type of one operand
+/// \tparam X2 the data type of the other operand
+template <typename X1, typename X2>
+struct get_binop_datatype {
+  using type = typename get_binop_datatype_impl<
+      typename upcast_if_derived_vector_type<X1>::type,
+      typename upcast_if_derived_vector_type<X2>::type>::type;
+
+  static_assert(
+      not std::is_same_v<type, std::bool_constant<false>>,
+      "You are attempting to perform a binary arithmetic operation between "
+      "two data types, but the data type of the result is not known within "
+      "TensorExpressions. See tenex::detail::get_binop_datatype_impl.");
+};
+
+/// \brief Whether or not it is permitted to perform binary arithmetic
+/// operations with the given types within `TensorExpression`
+///
+/// \details
+/// See `get_binop_datatype_impl` for which data type combinations have a
+/// defined result type
+///
+/// \tparam X1 the data type of one operand
+/// \tparam X2 the data type of the other operand
+template <typename X1, typename X2>
+struct binop_datatypes_are_supported {
+  using type = std::bool_constant<not std::is_same_v<
```

Comment:
After thinking about this more, I don't think I follow how this could inherit from `std::is_same` since I need to check if something is `not std::is_same`. The best I can think of is keeping the `std::bool_constant` and inheriting from that like:

```
template <typename X1, typename X2>
struct binop_datatypes_are_supported
    : std::bool_constant<not std::is_same_v<
          typename get_binop_datatype_impl<
              typename upcast_if_derived_vector_type<X1>::type,
              typename upcast_if_derived_vector_type<X2>::type>::type,
          NoSuchType>> {};
```

Did you have something different in mind that I'm not thinking of or is this fine?

Reply:
nah that's fine :)

Reply:
https://en.cppreference.com/w/cpp/types/negation

Reply:
Thanks, I haven't heard of this either, but I'll do the negation